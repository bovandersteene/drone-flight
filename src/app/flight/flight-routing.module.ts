import { NgModule } from "@angular/core";
import { RouterModule, Routes } from "@angular/router";
import { FlightComponent } from "./flight/flight.component";

const routes: Routes = [
  {
    path: "",
    component: FlightComponent
  },
  {
    path: ":id",
    component: FlightComponent
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class FlightRoutingModule {}
