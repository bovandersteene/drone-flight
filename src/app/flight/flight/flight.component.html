<form [formGroup]="flightForm" (submit)="saveFlight()">
  <mat-card>
    <mat-card-content>
      <app-map
        *ngIf="(flight$ | async) as flight"
        [latitude]="flight.location.lat"
        [longitude]="flight.location.lng"
        [paths]="flight.path"
        (changeLocation)="changeLocation($event)"
        (changePath)="changePath($event)"
      ></app-map>
    </mat-card-content>
    <mat-card-footer>
      <mat-form-field class="example-full-width">
        <input
          type="text"
          formControlName="name"
          matInput
          placeholder="Flight name"
        />
      </mat-form-field>
      <button mat-raised-button color="primary" type="submit">Save</button>
    </mat-card-footer>
  </mat-card>
</form>
