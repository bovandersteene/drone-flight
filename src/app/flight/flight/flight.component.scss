mat-card-footer {
  display: flex;
  padding: 0 20px 20px;
  mat-form-field {
    flex: 1;
  }
  button {
    margin: 10px 0 10px 20px;
  }
}

form {
  display: block;
  height: calc(100% - 40px);
  mat-card {
    height: calc(100% - 36px);
    mat-card-content {
      height: calc(100% - 80px);
    }
  }
}
