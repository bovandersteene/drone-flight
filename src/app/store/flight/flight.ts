import { LatLngLiteral } from "@agm/core";

export interface Flight {
  readonly id: number;
  readonly name: string;
  readonly location: LatLngLiteral;
  readonly path: Array<LatLngLiteral>;
}
