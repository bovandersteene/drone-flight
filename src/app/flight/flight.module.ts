import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";

import { FlightRoutingModule } from "./flight-routing.module";
import { FlightComponent } from "./flight/flight.component";
import { MapComponent } from "./map/map.component";
import { SharedModule } from "../shared/shared.module";
import { FormsModule, ReactiveFormsModule } from "@angular/forms";
import { AgmCoreModule } from "@agm/core";

@NgModule({
  declarations: [FlightComponent, MapComponent],
  imports: [
    CommonModule,
    FlightRoutingModule,
    SharedModule,
    ReactiveFormsModule,
    FormsModule,
    AgmCoreModule
  ]
})
export class FlightModule {}
