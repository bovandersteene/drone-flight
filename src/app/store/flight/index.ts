import { FEATURE_NAME } from "./flight.selector";
import { reducer } from "./flight.reducer";
import * as selectors from "./flight.selector";
export * from "./flight";

export const fromFlight = {
  FEATURE_NAME,
  reducer,
  selectors
};
