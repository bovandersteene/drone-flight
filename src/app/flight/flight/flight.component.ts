import { Component, OnDestroy, OnInit } from "@angular/core";
import { FormControl, FormGroup, Validators } from "@angular/forms";
import { select, Store } from "@ngrx/store";
import { FlightState } from "../../store/flight/flight.state";
import { Flight, fromFlight } from "../../store/flight";
import { SaveFlight } from "../../store/flight/flight.actions";
import { Observable, Subject } from "rxjs";
import { LatLngLiteral } from "@agm/core";
import { ActivatedRoute } from "@angular/router";
import { map, switchMap, tap } from "rxjs/operators";
import { getLocation } from "../../store/flight/flight.selector";

@Component({
  selector: "app-flight",
  templateUrl: "./flight.component.html",
  styleUrls: ["./flight.component.scss"]
})
export class FlightComponent implements OnInit, OnDestroy {
  flightForm: FormGroup;
  flight$: Observable<Flight>;
  location$: Observable<LatLngLiteral>;

  private destroy$ = new Subject();
  private changedPath: Array<LatLngLiteral>;

  constructor(
    private readonly store: Store<FlightState>,
    private readonly router: ActivatedRoute
  ) {}

  ngOnInit() {
    this.location$ = this.store.pipe(select(getLocation));

    this.flightForm = new FormGroup({
      id: new FormControl(undefined, [Validators.required]),
      name: new FormControl("", [Validators.required]),
      location: new FormControl(location, [Validators.required])
    });

    this.flight$ = this.router.params.pipe(
      map(param => param.id),
      switchMap(id =>
        this.store.pipe(select(fromFlight.selectors.findFlight(+id)))
      ),
      tap(({ name, location, id }) =>
        this.flightForm.setValue({ name, location, id })
      ),
      tap(({ path }) => this.changePath(path))
    );
  }

  saveFlight() {
    if (this.flightForm.valid && this.changedPath && this.changedPath.length) {
      const { id, name, location } = this.flightForm.value;
      const flight: Flight = {
        id,
        name,
        location,
        path: this.changedPath
      };
      this.store.dispatch(new SaveFlight(flight));
    }
  }

  changeLocation(coords: LatLngLiteral) {
    this.flightForm.get("location").setValue(coords);
  }

  changePath(path: Array<LatLngLiteral>) {
    this.changedPath = path;
  }

  ngOnDestroy() {
    this.destroy$.next();
  }
}
