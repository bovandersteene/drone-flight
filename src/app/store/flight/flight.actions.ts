import { Action } from "@ngrx/store";
import { Flight } from "./flight";
import { LatLngLiteral } from "@agm/core";

export enum FlightActionTypes {
  LoadFlights = "[Flight] Load Flights",
  LoadFlightsSuccess = "[Flight] Load Flights success",
  SaveFlight = "[Flight] Save Flight",
  SaveFlightSuccess = "[Flight] Save Flight success",
  ActiveLocation = "[Flight] Active location"
}

export class LoadFlights implements Action {
  readonly type = FlightActionTypes.LoadFlights;
}

export class LoadFlightsSuccess implements Action {
  readonly type = FlightActionTypes.LoadFlightsSuccess;
  readonly payload: { flights: Array<Flight> };

  constructor(flights: Array<Flight>) {
    this.payload = { flights };
  }
}

export class SaveFlight implements Action {
  readonly type = FlightActionTypes.SaveFlight;
  readonly payload: { flight: Flight };

  constructor(flight: Flight) {
    this.payload = { flight };
  }
}

export class SaveFlightSuccess implements Action {
  readonly type = FlightActionTypes.SaveFlightSuccess;
  readonly payload: { flight: Flight };

  constructor(flight: Flight) {
    this.payload = { flight };
  }
}

export class ActiveLocation implements Action {
  readonly type = FlightActionTypes.ActiveLocation;
  readonly payload: { location: LatLngLiteral };

  constructor(lat: number, lng: number) {
    this.payload = { location: { lat, lng } };
  }
}

export type FlightActions =
  | SaveFlight
  | ActiveLocation
  | SaveFlightSuccess
  | LoadFlights
  | LoadFlightsSuccess;
