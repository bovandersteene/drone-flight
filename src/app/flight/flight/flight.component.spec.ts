import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { FlightComponent } from "./flight.component";
import { NO_ERRORS_SCHEMA } from "@angular/core";
import { MockStoreModule } from "../../../../test-helper";
import { ActivatedRoute } from "@angular/router";
import { Store } from "@ngrx/store";
import { FlightState } from "../../store/flight/flight.state";
import { Flight } from "../../store/flight";
import { Subject } from "rxjs";
import { SaveFlight } from "../../store/flight/flight.actions";

const paramsSubject = new Subject();

class RouterMock {
  params = paramsSubject;
}

const location = { lat: 46.5188994165887, lng: 6.610714181232424 };
const flight: Flight = {
  id: 2,
  name: "Flight Lausanne",
  path: [
    { lat: 46.51952679624158, lng: 6.611394037382979 },
    { lat: 46.51926061859036, lng: 6.6125447888694 },
    { lat: 46.5184769874756, lng: 6.613104113268037 },
    { lat: 46.51821775486644, lng: 6.612684096100111 },
    { lat: 46.51739173123267, lng: 6.61203145980835 },
    { lat: 46.51803007072376, lng: 6.610243495444138 },
    { lat: 46.51767033135854, lng: 6.609494907671888 },
    { lat: 46.51865584543093, lng: 6.6081090018919895 },
    { lat: 46.51913747854242, lng: 6.609115291277249 }
  ],
  location: { lat: 46.5188994165887, lng: 6.610714181232424 }
};

describe("FlightComponent", () => {
  let component: FlightComponent;
  let fixture: ComponentFixture<FlightComponent>;
  let store: Store<FlightState>;
  let router: ActivatedRoute;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [FlightComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        {
          provide: ActivatedRoute,
          useClass: RouterMock
        }
      ],
      imports: [
        MockStoreModule.forRoot("flight", {
          flights: [flight],
          activeLocation: location
        })
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(FlightComponent);
    router = TestBed.get(ActivatedRoute);
    store = TestBed.get(Store);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });
  it("should create", () => {
    expect(component).toBeTruthy();
  });

  describe("navigate to an existing flight", () => {
    beforeEach(() => {
      jest.spyOn(component, "changePath");
      paramsSubject.next({ id: flight.id });
    });
    it("should set the form", () => {
      const formValue = component.flightForm.value;
      expect(formValue.id).toBe(flight.id);
      expect(formValue.name).toBe(flight.name);
      expect(formValue.location).toBe(flight.location);
      expect(component.changePath).toHaveBeenCalledWith(flight.path);
    });
    it("should render the component with values", () => {
      fixture.detectChanges();
      expect(fixture.debugElement.nativeElement).toMatchSnapshot(
        "existing flight"
      );
    });
  });
  describe("navigate to an NON existing flight", () => {
    beforeEach(() => {
      jest.spyOn(component, "changePath");
      paramsSubject.next({});
    });
    it("should set the form with default values", () => {
      const formValue = component.flightForm.value;
      expect(formValue.id).not.toBe(flight.id);
      expect(formValue.id).toBeDefined();
      expect(formValue.location).toBe(location);
      expect(formValue.name.length).toBe(0);
      expect(component.changePath).toHaveBeenCalledWith([]);
    });
    it("should render the component with values", () => {
      fixture.detectChanges();
      expect(fixture.debugElement.nativeElement).toMatchSnapshot(
        "NON existing flight"
      );
    });
  });

  describe("saveFlight", () => {
    describe("the form is not valid", () => {
      it("should not dispatch the saveFlight", () => {
        jest.spyOn(store, "dispatch");
        component.flightForm.setValue({ id: 123, name: "", location: {} });
        component.saveFlight();
        expect(store.dispatch).not.toHaveBeenCalled();
      });
    });
    describe("the form is valid", () => {
      describe("the paths are empty", () => {
        it("should not dispatch the saveFlight", () => {
          jest.spyOn(store, "dispatch");
          component.flightForm.setValue({
            id: 123,
            name: "new flight",
            location
          });
          expect(component.flightForm.valid).toBe(true);
          component.changePath([]);
          component.saveFlight();
          expect(store.dispatch).not.toHaveBeenCalled();
        });
        describe("the paths is set", () => {
          it("should not dispatch the saveFlight", () => {
            jest.spyOn(store, "dispatch");
            const { id, name } = flight;
            component.flightForm.setValue({
              id,
              name,
              location: flight.location
            });
            expect(component.flightForm.valid).toBe(true);
            component.changePath(flight.path);
            component.saveFlight();
            expect(store.dispatch).toHaveBeenCalledWith(new SaveFlight(flight));
          });
        });
      });
    });

    describe("changeLocation", () => {
      it("should set the location", () => {
        component.flightForm.get("location").setValue(undefined);
        component.changeLocation(location);
        expect(component.flightForm.value.location).toBe(location);
      });
    });
  });
});
