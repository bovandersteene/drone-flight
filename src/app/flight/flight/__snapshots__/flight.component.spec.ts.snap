// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightComponent navigate to an NON existing flight should render the component with values: NON existing flight 1`] = `
<div
  id="root4"
  ng-version="7.1.0"
>
  <form>
    <mat-card>
      <mat-card-content>
        <!--bindings={
  "ng-reflect-ng-if": "[object Object]"
}-->
        <app-map />
      </mat-card-content>
      <mat-card-footer>
        <mat-form-field
          class="example-full-width"
        >
          <input
            formcontrolname="name"
            matinput=""
            placeholder="Flight name"
            type="text"
          />
        </mat-form-field>
        <button
          color="primary"
          mat-raised-button=""
          type="submit"
        >
          Save
        </button>
      </mat-card-footer>
    </mat-card>
  </form>
</div>
`;

exports[`FlightComponent navigate to an existing flight should render the component with values: existing flight 1`] = `
<div
  id="root2"
  ng-version="7.1.0"
>
  <form>
    <mat-card>
      <mat-card-content>
        <!--bindings={
  "ng-reflect-ng-if": "[object Object]"
}-->
        <app-map />
      </mat-card-content>
      <mat-card-footer>
        <mat-form-field
          class="example-full-width"
        >
          <input
            formcontrolname="name"
            matinput=""
            placeholder="Flight name"
            type="text"
          />
        </mat-form-field>
        <button
          color="primary"
          mat-raised-button=""
          type="submit"
        >
          Save
        </button>
      </mat-card-footer>
    </mat-card>
  </form>
</div>
`;
